disperser/controller/encoding_manager.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	for _, blob := range blobMetadatas {
+		blobKey, err := blob.BlobHeader.BlobKey()
+		if err != nil {
+			e.logger.Error("failed to get blob key", "err", err, "requestedAt", blob.RequestedAt, "paymentMetadata", blob.BlobHeader.PaymentMetadata)
+			continue
+		}
+		e.lastUpdatedAt = blob.UpdatedAt
+
+		// Encode the blobs
+		e.pool.Submit(func() {
+			for i := 0; i < e.NumEncodingRetries+1; i++ {
+				encodingCtx, cancel := context.WithTimeout(ctx, e.EncodingRequestTimeout)
+				fragmentInfo, err := e.encodeBlob(encodingCtx, blobKey, blob)
+				cancel()
+				if err != nil {
+					e.logger.Error("failed to encode blob", "blobKey", blobKey.Hex(), "err", err)
+					continue
+				}
+				relayKeys, err := GetRelayKeys(e.NumRelayAssignment, e.AvailableRelays)
+				if err != nil {
+					e.logger.Error("failed to get relay keys", "err", err)
+					// Stop retrying
+					break
+				}
+				cert := &corev2.BlobCertificate{
+					BlobHeader:           blob.BlobHeader,
+					ReferenceBlockNumber: uint64(currentBlockNumber),
+					RelayKeys:            relayKeys,
+				}
+
+				storeCtx, cancel := context.WithTimeout(ctx, e.StoreTimeout)
+				err = e.blobMetadataStore.PutBlobCertificate(storeCtx, cert)
+				cancel()
+				if err != nil && !errors.Is(err, dispcommon.ErrAlreadyExists) {
+					e.logger.Error("failed to put blob certificate", "err", err)
+					continue
+				}
+
+				storeCtx, cancel = context.WithTimeout(ctx, e.StoreTimeout)
+				err = e.blobMetadataStore.MarkBlobEncoded(storeCtx, blobKey, fragmentInfo)
+				cancel()
+				if err == nil || errors.Is(err, dispcommon.ErrAlreadyExists) {
+					// Successfully updated the status to Encoded
+					return
+				}
+
+				e.logger.Error("failed to update blob status to Encoded", "blobKey", blobKey.Hex(), "err", err)
+				time.Sleep(time.Duration(math.Pow(2, float64(i))) * time.Second) // Wait before retrying
+			}
+
+			storeCtx, cancel := context.WithTimeout(ctx, e.StoreTimeout)
+			err = e.blobMetadataStore.MarkBlobFailed(storeCtx, blobKey)
+			cancel()
+			if err != nil {
+				e.logger.Error("failed to update blob status to Failed", "blobKey", blobKey.Hex(), "err", err)
+				return
+			}
+		})
+	}
+
+	return nil
+}
+
+func (e *EncodingManager) encodeBlob(ctx context.Context, blobKey corev2.BlobKey, blob *v2.BlobMetadata) (*encoding.FragmentInfo, error) {
+	encodingParams, err := blob.BlobHeader.GetEncodingParams()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get encoding params: %w", err)
+	}
+	return e.encodingClient.EncodeBlob(ctx, blobKey, encodingParams)
+}
+
+func GetRelayKeys(numAssignment uint16, availableRelays []corev2.RelayKey) ([]corev2.RelayKey, error) {
+	if int(numAssignment) > len(availableRelays) {
+		return nil, fmt.Errorf("numAssignment (%d) cannot be greater than numRelays (%d)", numAssignment, len(availableRelays))
+	}
+	relayKeys := availableRelays
+	// shuffle relay keys
+	for i := len(relayKeys) - 1; i > 0; i-- {
+		j := rand.Intn(i + 1)
```

Comment:
should this be cryptographically secure randomness? If so `crypto/rand` should be used.

Reply:
I don't think so. It just has to be random enough. Performance is probably more important than security